app/integrations.go

```diff
@@ -0,0 +1,170 @@
+package app
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v6/app/request"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+type ListedApp struct {
+	Manifest struct {
+		AppID       string `json:"app_id"`
+		DisplayName string `json:"display_name"`
+		Version     string `json:"version"`
+	}
+
+	Installed bool `json:"installed"`
+	Enabled   bool `json:"enabled"`
+}
+
+func (a *App) GetInstalledIntegrations() ([]*model.InstalledIntegration, *model.AppError) {
+	return a.ch.getInstalledIntegrations()
+}
+
+func (ch *Channels) getInstalledIntegrations() ([]*model.InstalledIntegration, *model.AppError) {
+	out := []*model.InstalledIntegration{}
+
+	pluginsEnvironment := ch.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return out, nil
+	}
+
+	plugins, err := pluginsEnvironment.Available()
+	if err != nil {
+		return nil, model.NewAppError("getInstalledIntegrations", "app.plugin.sync.read_local_folder.app_error", nil, err.Error(), 0)
+	}
+
+	for _, p := range plugins {
+		ignore := false
+		for _, id := range model.InstalledIntegrationsIgnoredPlugins {
+			if p.Manifest.Id == id {
+				ignore = true
+				break
+			}
+		}
+
+		if !ignore {
+			integration := &model.InstalledIntegration{
+				Type:    "plugin",
+				ID:      p.Manifest.Id,
+				Name:    p.Manifest.Name,
+				Version: p.Manifest.Version,
+				Enabled: pluginsEnvironment.IsActive(p.Manifest.Id),
+			}
+
+			out = append(out, integration)
+		}
+	}
+
+	if pluginsEnvironment.IsActive("com.mattermost.apps") {
+		enabledApps, appErr := ch.getInstalledApps()
+		if appErr != nil {
+			ch.srv.Log.Warn("Failed to fetch installed Apps", mlog.Err(appErr))
+			enabledApps = []ListedApp{}
+		}
+
+		for _, ap := range enabledApps {
```

Comment:
Can't use the same set trick down here (unless we build a set of the currently added integrations, but not sure I like that). But we could use a [labeled loop](https://www.ardanlabs.com/blog/2013/11/label-breaks-in-go.html).

```go
	AppsLoop:
		for _, ap := range enabledApps {
			for _, integration := range out {
				if integration.ID == ap.Manifest.AppID {
					integration.Type = "plugin-app"
					continue AppsLoop
				}
			}

			integration := &model.InstalledIntegration{
				Type:    "app",
				ID:      ap.Manifest.AppID,
				Name:    ap.Manifest.DisplayName,
				Version: ap.Manifest.Version,
				Enabled: ap.Enabled,
			}

			out = append(out, integration)
		}
```